gpflow/base.py

```diff
@@ -28,14 +29,20 @@
 
 DType = Union[np.dtype, tf.DType]
 
-# It would be nice to use something more interesting than `Any` here, but it looks like the
-# infrastructure in the rest of the ecosystem isn't really set up for this yet. Maybe when we
-# get Python 3.11?
 if TYPE_CHECKING:  # pragma: no cover
-    AnyNDArray = Union["np.ndarray[Any, Any]"]
-elif Version(np.__version__) < Version("1.20.0"):
+    # It would be nice to use something more interesting than `Any` here, but valid values depend on
+    # the version of `numpy`, and `mypy` doesn't support branching on that. Maybe, when we bump
+    # minimal versions of Python and TensorFlow to > 3.7 and 2.4, we can set a minimum version of
+    # numpy to >= 1.21, and we can get rid of this.
+    AnyNDArray = Any  # Valid for all versions of numpy.
+    # AnyNDArray = Union["np.ndarray"]  # Valid for numpy < 1.21.0
```

Comment:
I think the following workaround works:

```python
if TYPE_CHECKING:
    GENERIC_NDARRAYS = True
    if GENERIC_NDARRAYS:
        AnyNDArray = Union["np.ndarray[Any, Any]"]
    else:
        AnyNDArray = Union["np.ndarray"] # type: ignore
```

then for numpy<1.21.0 you call mypy with `--always-false GENERIC_NDARRAYS` to get the second definition.